File: Simulations/StepperMotorWithInterrupts/StepperMotorSpeedProfile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace MotorShieldSim
{
    //*****************************************************************************************
    //
    // SpeedProfileSegment
    //
    class SpeedProfileSegment
    {
      // specified by caller
        public float  startTime; // seconds
        public float  duration;  // seconds

        public float  finalRPM;
        public float  initialRPM; // final RPM of previous segment

      // calculated
        public UInt16 initialPPS; // final PPS of previous segment
        public UInt16 finalPPS;   // calculated from caller specified final RPM

        public float C1;  // see Visio drawing or Matlab stepper.m
        public float C2;

        public UInt16 numberSteps; // calculated number of steps this segment
    }

    public class StepperMotorSpeedProfile : MotorProfile
    {
        private List<SpeedProfileSegment> speedProfile = new List<SpeedProfileSegment> ();

        private int getIndex = 0;    // used to read-back profile, index of current segment in speedProfile[]

        private UInt16 stepsSent = 0;   // number of steps that have been sent for current profile segment

        private float profileStartTime = 0;


        public UInt16 NumberSteps
        {
            get
            {
                UInt16 count = 0;

                foreach (SpeedProfileSegment seg in speedProfile)
                    count += seg.numberSteps;

                return count;
            }
        }



        public ushort StepsPerRevolution {get; set;} // motor characteristic
        public MotorDirection Direction {get; set;}



        //********************************************************************************
        //
        // Constructor
        //
        public StepperMotorSpeedProfile (ushort stepsPerRev, MotorDirection dir)
        {
            StepsPerRevolution = stepsPerRev;
            Direction = dir;
        }

        //********************************************************************************
        //
        // StartAt ()
        //
        public void StartAt (float time)
        {
            profileStartTime = time;
        }

        //********************************************************************************
        //
        // Ramp - add a speed change to profile
        //
        public void Ramp (float seconds, 
                          float finalRPM)
        {
            SpeedProfileSegment segment = new SpeedProfileSegment ();
             
            if (speedProfile.Count == 0)
            {
                segment.initialRPM = 0;
                segment.initialPPS = 0;
                segment.startTime  = 0;
            }
            else
            {
                SpeedProfileSegment previous = speedProfile [speedProfile.Count - 1];

                segment.initialRPM = previous.finalRPM;
                segment.initialPPS = previous.finalPPS;
                segment.startTime  = previous.startTime + previous.duration;
            }

            segment.finalRPM = finalRPM;
            segment.finalPPS = (ushort) (segment.finalRPM  * StepsPerRevolution * Constants.MinutesPerSecond);
            segment.duration = seconds;

            segment.numberSteps = (ushort) (Math.Min (segment.initialPPS, segment.finalPPS) * segment.duration + 0.5 * segment.duration * Math.Abs (segment.finalPPS - segment.initialPPS));

            segment.C1 = segment.initialPPS;
            segment.C2 = 0.5f * (segment.finalPPS - segment.initialPPS) / segment.duration;

            speedProfile.Add (segment);
        }

        //********************************************************************************
        //
        // Hold - add a constant speed section to profile
        //
        public void Hold (float seconds)
        {
            if  (speedProfile.Count == 0)
                Ramp (seconds, 0);
            else
                Ramp (seconds, speedProfile [speedProfile.Count - 1].finalRPM);
        }

        //********************************************************************************
        //
        // ResetReadback () - called prior to beginning a read-back
        //
        public void ResetReadback ()
        {
            getIndex = 0;
            stepsSent = 0;
        }

        //********************************************************************************
        //
        // GetNext - get time of next pulse. returns true if this is last value
        //
        public bool GetNext (ref float nextTime)
        {
            SpeedProfileSegment segment = speedProfile [getIndex];

            stepsSent++; // calculate the time of the next step

          // if in a speed change section
            if (Math.Abs (segment.C2) > 1e-4) // arbitrary small threshold
            {
                nextTime = (float) (-segment.C1 + Math.Sqrt (segment.C1 * segment.C1 - 4 * segment.C2 * -stepsSent)) / (2 * segment.C2);
            }
            else // constant speed
            {
                nextTime = stepsSent / segment.C1;
            }

            nextTime += segment.startTime;
            nextTime += profileStartTime;

          // see if we're at the end of a segment. if so, go to next if any remaining
            if (stepsSent == speedProfile [getIndex].numberSteps)
            {
                getIndex++;
                stepsSent = 0;
            }

            return getIndex == speedProfile.Count;
        }
    }
}

File: Play/Archive/MotorsAndEncoders_ORIGINAL/StepperMotorSpeedProfile.cpp

// StepperMotorSpeedProfile.cpp

#include "Arduino.h"
#include "Utils.h"
#include "StepperMotorSpeedProfile.h"

static float MinutesPerSecond = (float) (1 / 60.0);

//********************************************************************************
//
// Constructor
//
StepperMotorSpeedProfile::StepperMotorSpeedProfile ()
{

}

//********************************************************************************
//
// Initialize
//
bool StepperMotorSpeedProfile::Initialize (unsigned short stepsPerRev, MotorDirection dir)
{
    ResetReadback ();
    stepsPerRevolution = stepsPerRev;
    direction = dir;
    StartAt (0);
    
    return true;
}
    
//********************************************************************************
//
// StartAt ()
//
void StepperMotorSpeedProfile::StartAt (float time)
{
    profileStartTime = time;   
}

//********************************************************************************

void StepperMotorSpeedProfile::Print ()
{
    char obuf [100];
    char nbuf [100];
    
    sprintf (obuf, "SpeedProfile has %d segments", putIndex); Serial.println (obuf);

    for (int i=0; i<putIndex; i++)
    {
        StepperSpeedProfileSegment& segment = speedProfile [i];

        Serial.print ("Segment: "); Serial.println (i);
        
        sprintf (obuf, "StartTime %s", Utils::FloatToString (segment.startTime, nbuf)); Serial.println (obuf);        
        sprintf (obuf, "Duration %s",  Utils::FloatToString (segment.duration,  nbuf)); Serial.println (obuf);

        sprintf (obuf, "initial RPM %s", Utils::FloatToString (segment.initialRPM, nbuf)); Serial.println (obuf);
        sprintf (obuf, "final RPM %s",   Utils::FloatToString (segment.finalRPM,   nbuf)); Serial.println (obuf);

        /**
        
        sprintf (obuf, "initial PPS %d", segment.initialPPS); Serial.println (obuf);
        sprintf (obuf, "final PPS %d",   segment.finalPPS);   Serial.println (obuf);
        
        sprintf (obuf, "C1 %s", Utils::FloatToString (segment.C1); Serial.println (obuf);
        sprintf (obuf, "C2 %s", Utils::FloatToString (segment.C2); Serial.println (obuf);
        
        sprintf (obuf, "number steps %d", segment.numberSteps); Serial.println (obuf);      
        **/
        Serial.println (" ");
    }
}

//********************************************************************************
//
// Ramp - add a speed change to profile
//
void StepperMotorSpeedProfile::Ramp (float seconds, float finalRPM)
{    
    if (putIndex >= MaxSpeedProfileSegment)
        return;
        
    StepperSpeedProfileSegment& segment = speedProfile [putIndex]; // note this is a reference

    if (putIndex == 0)    
    {
        segment.initialRPM = 0;
        segment.initialPPS = 0;
        segment.startTime  = 0;
    }
    else
    {    
        StepperSpeedProfileSegment& previous = speedProfile [putIndex - 1];

        segment.initialRPM = previous.finalRPM;
        segment.initialPPS = previous.finalPPS;
        segment.startTime  = previous.startTime + previous.duration;
    }

    segment.finalRPM = finalRPM;
    segment.finalPPS = segment.finalRPM  * stepsPerRevolution * MinutesPerSecond;
    segment.duration = seconds;

    segment.numberSteps = min (segment.initialPPS, segment.finalPPS) * segment.duration + 0.5 * segment.duration * fabs (segment.finalPPS - segment.initialPPS);

    segment.C1 = segment.initialPPS;
    segment.C2 = 0.5f * (segment.finalPPS - segment.initialPPS) / segment.duration;

    putIndex++;
}

//********************************************************************************
//
// Hold - add a constant speed section to profile
//
void StepperMotorSpeedProfile::Hold (float seconds)
{
    if  (putIndex == 0)
        Ramp (seconds, 0);
    else
        Ramp (seconds, speedProfile [putIndex - 1].finalRPM);
}

//********************************************************************************
//
// ResetReadback () - called prior to beginning a read-back
//
void StepperMotorSpeedProfile::ResetReadback ()
{
    getIndex = 0;
    stepsSent = 0;
}

//********************************************************************************
//
// GetNext - get time of next pulse. returns true if this is last value
//
bool StepperMotorSpeedProfile::GetNext (float& nextTime)
{
    StepperSpeedProfileSegment& segment = speedProfile [getIndex];

    stepsSent++; // calculate the time of the next step

  // if in a speed change section
    if (fabs (segment.C2) > 1e-4) // arbitrary small threshold
    {
      nextTime = (float) (-segment.C1 + sqrt (segment.C1 * segment.C1 + 4 * segment.C2 * stepsSent)) / (2 * segment.C2);
    }
    else // constant speed
    {
        nextTime = stepsSent / segment.C1;
    }

    nextTime += segment.startTime;
    nextTime += profileStartTime;

  // see if we're at the end of a segment. if so, go to next if any remaining
    if (stepsSent == speedProfile [getIndex].numberSteps)
    {
        getIndex++;
        stepsSent = 0;

        Serial.println ("Segment done");
    }

    return getIndex == putIndex;
}
